src/wallet/walletdb.cpp

```diff
@@ -664,6 +664,12 @@ ReadKeyValue(CWallet* pwallet, DataStream& ssKey, CDataStream& ssValue,
                 wss.m_descriptor_caches[id] = DescriptorCache();
             }
             pwallet->LoadDescriptorScriptPubKeyMan(id, desc);
+
+            // Prior to doing anything with this spkm, verify ID calculation is always the same.
+            if (id != pwallet->GetDescriptorScriptPubKeyMan(desc)->GetID()) {
+                wss.descriptor_unknown = true;
+                return error("The descriptor ID calculated by the wallet differs from the one in DB");
```

Comment:
In 080ecd855bc4194396c58e71677a3e50c7b0abff "wallet: do not allow loading descriptor with an invalid ID"

The pattern we use for errors in `ReadKeyValue` is to assign the `strErr` rather than `return error("...")`.

```suggestion
                strErr = "The descriptor ID calculated by the wallet differs from the one in DB");
                return false
```

Reply:
Yeah, I did it on purpose, so the error is actually logged. The ´descriptor_unknown´ field overwrites ´strErr´ with another message.

But, what could do is actually change the ´descriptor_unknown´ logging so it appends the error at the end (it will also log the currently not logged  ´std::ios_base::failure´ error that is above).

Reply:
I'm not sure that `descriptor_unknown` needs to be set here. The descriptor isn't really unknown, just our computed id doesn't match the stored one. This is more on the level of a critical corruption error, in the same way that invalid keys are.

Error handling here should be a lot easier with #24914.

Reply:
> I'm not sure that descriptor_unknown needs to be set here. The descriptor isn't really unknown, just our computed id doesn't match the stored one. This is more on the level of a critical corruption error, in the same way that invalid keys are.

Yeah, but we don't have a way to set a corruption error right now. If the function only return false with the strErr, it will be tagged as a non-critical error. Thus why re-used the `descriptor_unknown` flag.

> Error handling here should be a lot easier with https://github.com/bitcoin/bitcoin/pull/24914.

yessss. Thus why I didn't care much on adding more code to it. While it works, should be fine.